tests/test-ray-tracing-hitobject-intrinsics.cpp

```diff
@@ -0,0 +1,646 @@
+        const TestResultCudaAligned* result =
+            reinterpret_cast<const TestResultCudaAligned*>(resultBlob->getBufferPointer());
+        CHECK_EQ(result->rayOrigin[0], 0.1f);
+        CHECK_EQ(result->rayOrigin[1], 0.1f);
+        CHECK_EQ(result->rayOrigin[2], 0.1f);
+    }
+    else
+    {
+        REQUIRE_CALL(device->readBuffer(test.resultBuffer, 0, sizeof(TestResult), resultBlob.writeRef()));
+        const TestResult* result = reinterpret_cast<const TestResult*>(resultBlob->getBufferPointer());
+        CHECK_EQ(result->rayOrigin[0], 0.1f);
+        CHECK_EQ(result->rayOrigin[1], 0.1f);
+        CHECK_EQ(result->rayOrigin[2], 0.1f);
+    }
+}
+
+// Disabled under CUDA/OptiX and D3D12 due to https://github.com/shader-slang/slang/issues/8615
+GPU_TEST_CASE("ray-tracing-hitobject-query-hit-ray-object-direction", ALL & ~CUDA & ~D3D12)
+{
+    if (!device->hasFeature(Feature::RayTracing))
+        SKIP("ray tracing not supported");
+    if (!device->hasFeature(Feature::ShaderExecutionReordering))
+        SKIP("shader execution reordering not supported");
+
+    RayTracingHitObjectIntrinsicsTest test;
+    test.init(device);
+    test.run(
+        "test-ray-tracing-hitobject-intrinsics",
+        "rayGenShaderQueryRayObjectDirection",
+        "closestHitNOP",
+        "missNOP"
+    );
+
+    ComPtr<ISlangBlob> resultBlob;
+
+    if (device->getDeviceType() == DeviceType::CUDA)
+    {
+        REQUIRE_CALL(device->readBuffer(test.resultBuffer, 0, sizeof(TestResultCudaAligned), resultBlob.writeRef()));
+        const TestResultCudaAligned* result =
+            reinterpret_cast<const TestResultCudaAligned*>(resultBlob->getBufferPointer());
+        CHECK_EQ(result->rayDirection[0], 0.0f);
+        CHECK_EQ(result->rayDirection[1], 0.0f);
+        CHECK_EQ(result->rayDirection[2], 1.0f);
+    }
+    else
+    {
+        REQUIRE_CALL(device->readBuffer(test.resultBuffer, 0, sizeof(TestResult), resultBlob.writeRef()));
+        const TestResult* result = reinterpret_cast<const TestResult*>(resultBlob->getBufferPointer());
+        CHECK_EQ(result->rayDirection[0], 0.0f);
+        CHECK_EQ(result->rayDirection[1], 0.0f);
+        CHECK_EQ(result->rayDirection[2], 1.0f);
+    }
+}
+
+GPU_TEST_CASE("ray-tracing-hitobject-make-hit", ALL | DontCreateDevice)
+{
+    // Limit the shader model to SM 6.6 for this test, since the NVAPI headers don't support MakeHit
+    // for newer shader models.
+    DeviceExtraOptions extraOptions;
+    extraOptions.d3d12HighestShaderModel = 0x66; // SM 6.6
+    device = createTestingDevice(ctx, ctx->deviceType, false, &extraOptions);
+    REQUIRE(device);
+
+    if (!device->hasFeature(Feature::RayTracing))
+        SKIP("ray tracing not supported");
+    if (!device->hasFeature(Feature::ShaderExecutionReordering))
+        SKIP("shader execution reordering not supported");
+
+
+    // Disabled under pre OptiX 9.0 due to https://github.com/shader-slang/slang/issues/8723
```

Comment:
There is at least one HitObject API (setting sbt index) that is not supported in OptiX 8.1, which otherwise supports most of the SER HitObject API. So the question will come if we just require OptiX 9.0 for SER.